core/miner/storage_fsm/impl/precommit_batcher_impl.cpp

```diff
@@ -0,0 +1,146 @@
+        std::make_shared<PreCommitBatcherImpl>(
+            max_wait, std::move(api), miner_address, max_wait);
+
+    batcher->cutoff_start_ = std::chrono::system_clock::now();
+    batcher->logger_ = common::createLogger("batcher");
+    batcher->logger_->info("Bather have been started");
+    batcher->handle_ = scheduler->schedule(max_wait, [=]() {
+      batcher->sendBatch();
+      batcher->handle_.reschedule(max_wait);  // TODO: maybe in gsl::finally
+    });
+    return batcher;
+  }
+
+  void PreCommitBatcherImpl::sendBatch() {
+    std::unique_lock<std::mutex> locker(mutex_, std::defer_lock);
+    // TODO(Elestrias): [FIL-398] goodFunds = mutualDeposit + MaxFee; -  for
+    // checking payable
+
+    if (batch_storage_.size() != 0) {
+      logger_->info("Sending procedure started");
+      auto head = api_->ChainHead().value();
+      auto minfo = api_->StateMinerInfo(miner_address_, head->key).value();
+
+      PreCommitBatch::Params params = {};
+      for (const auto &data : batch_storage_) {
+        mutual_deposit_ += data.second.deposit;
+        params.sectors.push_back(data.second.precommit_info);
+      }
+      auto encodedParams = codec::cbor::encode(params);
+      if (encodedParams.has_error()) {
+        logger_->error("Error has occurred during parameters encoding: {}",
+                       encodedParams.error().message());
+      }
+      logger_->info("Sending data to the network...");
+      auto maybe_signed = api_->MpoolPushMessage(
+          vm::message::UnsignedMessage(
+              miner_address_,
+              minfo.worker,  // TODO: handle worker
+              0,
+              mutual_deposit_,
+              {},
+              {},
+              25,  // TODO (m.tagirov) Miner actor v5 PreCommitBatch
+              MethodParams{encodedParams.value()}),
+          kPushNoSpec);
+
+      if (maybe_signed.has_error()) {
+        logger_->error("Error has occurred during batch sending: {}",
+                       maybe_signed.error().message());  // TODO: maybe logs
+      }
+
+      mutual_deposit_ = 0;
+      batch_storage_.clear();
+    }
+    cutoff_start_ = std::chrono::system_clock::now();
+    logger_->info("Sending procedure completed");
+  }
+
+  outcome::result<void> PreCommitBatcherImpl::forceSend() {
+    sendBatch();
+    handle_.reschedule(max_delay_);
+    return outcome::success();
+  }
+
+  outcome::result<void> PreCommitBatcherImpl::setPreCommitCutoff(
+      const ChainEpoch &curEpoch, const SectorInfo &si) {
+    ChainEpoch cutoffEpoch =
+        si.ticket_epoch
+        + static_cast<int64_t>(fc::kEpochsInDay + kChainFinality);
+    ChainEpoch startEpoch{};
+    for (auto p : si.pieces) {
+      if (!p.deal_info) {
+        continue;
+      }
+      startEpoch = p.deal_info->deal_schedule.start_epoch;
+      if (startEpoch < cutoffEpoch) {
+        cutoffEpoch = startEpoch;
+      }
+    }
+    if (cutoffEpoch <= curEpoch) {
+      OUTCOME_TRY(forceSend());
```

Comment:
Don't forget to remove OUTCOME_TRY after making forceSend() be just void.
And in this case you need to change
`outcome::result<void> PreCommitBatcherImpl::setPreCommitCutoff`
to
`void PreCommitBatcherImpl::setPreCommitCutoff`